pkg/eestream/encode.go

```diff
@@ -38,71 +38,103 @@ type ErasureScheme interface {
 }
 
 type encodedReader struct {
-	r               io.Reader
-	es              ErasureScheme
-	cv              *sync.Cond
-	inbuf           []byte
-	outbufs         [][]byte
-	piecesRemaining int
-	err             error
+	r       io.Reader
+	es      ErasureScheme
+	inbuf   []byte
+	outbufs [][]byte
+	chans   map[int]chan []byte
+	err     error
 }
 
 // EncodeReader will take a Reader and an ErasureScheme and return a slice of
 // Readers
-func EncodeReader(r io.Reader, es ErasureScheme) []io.Reader {
+func EncodeReader(r io.Reader, es ErasureScheme, maxBufferMemory int) []io.Reader {
 	er := &encodedReader{
 		r:       r,
 		es:      es,
-		cv:      sync.NewCond(&sync.Mutex{}),
 		inbuf:   make([]byte, es.DecodedBlockSize()),
 		outbufs: make([][]byte, es.TotalCount()),
+		chans:   make(map[int]chan []byte, es.TotalCount()),
 	}
 	readers := make([]io.Reader, 0, es.TotalCount())
 	for i := 0; i < es.TotalCount(); i++ {
-		er.outbufs[i] = make([]byte, 0, es.EncodedBlockSize())
 		readers = append(readers, &encodedPiece{
 			er: er,
 			i:  i,
 		})
 	}
+	if maxBufferMemory < 0 {
+		er.err = Error.New("negative max buffer memory")
+		return readers
+	}
+	chanSize := maxBufferMemory / (es.TotalCount() * es.EncodedBlockSize())
+	if chanSize < 1 {
+		chanSize = 1
+	}
+	for i := 0; i < es.TotalCount(); i++ {
+		er.chans[i] = make(chan []byte, chanSize)
+	}
+	go er.fillBuffer()
 	return readers
 }
 
-func (er *encodedReader) wait() (err error) {
-	// have we already failed? just return that
-	if er.err != nil {
-		return er.err
+func (er *encodedReader) fillBuffer() {
+	defer er.closeChannels()
+	for {
+		_, err := io.ReadFull(er.r, er.inbuf)
+		if err != nil {
+			return
+		}
+		err = er.es.Encode(er.inbuf, er.addToReader)
+		if err != nil {
+			return
+		}
+	}
+}
+
+func (er *encodedReader) addToReader(num int, data []byte) {
+	if er.chans[num] == nil {
+		// this channel is already closed for slowliness - skip it
+		return
 	}
-	// are other pieces still using buffer? wait on a condition variable for
-	// the last remaining piece to fill all the buffers.
-	if er.piecesRemaining > 0 {
-		er.cv.Wait()
-		// whoever broadcast a wakeup either set an error or filled the buffers.
-		// er.err might be nil, which means the buffers are filled.
-		return er.err
+	// data is reused by infecious, so add a copy to the channel
+	tmp := make([]byte, len(data))
+	copy(tmp, data)
+	// add data copy to the respective reader's channel
+	for {
+		select {
+		case er.chans[num] <- tmp:
+			return
+		// block for no more than 50 ms
+		case <-time.After(50 * time.Millisecond):
+			if er.isSlowChannel(num) {
+				close(er.chans[num])
+				er.chans[num] = nil
+				return
+			}
+		}
 	}
+}
 
-	// we are going to set an error or fill the buffers
-	defer er.cv.Broadcast()
-	defer func() {
-		// at the end of this function, if we're returning an error, set er.err
-		if err != nil {
-			er.err = err
+func (er *encodedReader) isSlowChannel(num int) bool {
```

Comment:
are you using `num` ?

Reply:
No :) This method actually checks if there are enough fast readers (i.e. their buffered channels are empty), so the caller of this method can treat the current channel as slow. This method requires more work and I'll clean it as part of this.